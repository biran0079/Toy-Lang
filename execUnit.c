#include "execUnit.h"

ExecUnit* newExecUnit(Value* ev, Node* p){
  ExecUnit* res = MALLOC(ExecUnit);
  res->ev = ev;
  res->p = p;
  return res;
}

void freeExecUnit(ExecUnit* eu){
  free(eu);
}

Value* exec(ExecUnit* eu){
  return eval(eu->ev, eu->p);
}
