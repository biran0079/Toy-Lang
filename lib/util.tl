fun getRandomList(n){
  res=[];
  while(n>0){
    res += rand();  
    n = n - 1;
  }
  return res;
}


import list;

fun internalQsort(a, left, right){
  if(left>=right)return;
  i=left;
  j=left+1;
  k=j;
  while(j<=right){
    if(a[j]<a[i]) {
      list.swap(a,k,j);
      k=k+1;
    }
    j=j+1;  
  }
  k=k-1;
  list.swap(a, left, k);
  internalQsort(a,left,k-1);
  internalQsort(a,k+1,right);
}
fun qsort(l){
  internalQsort(l,0,len(l)-1);  
}


fun merge(l1, l2){
  res=[];
  len1 = len(l1);
  len2 = len(l2);
  i=0;
  j=0;
  while(i<len1 && j<len2){
    if(l1[i]<l2[j]){
      res += l1[i];  
      i += 1;
    }else{
      res += l2[j];
      j = j+1;
    }
  }
  while(i<len1) {
    res += l1[i];  
    i += 1;
  }
  while(j<len2) {
    res += l2[j];
    j += 1;
  }
  return res;
}
fun msort(lst){
  length = len(lst);
  if(length<=1)return;
  mid = length / 2;
  l1 = list.prefix(lst, mid);
  l2 = list.suffix(lst, mid);
  msort(l1);
  msort(l2);
  t = merge(l1, l2);
  i=0;
  while(i<length){
    lst[i] = t[i];
    i=i+1;  
  }
}

