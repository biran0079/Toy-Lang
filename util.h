#ifndef _UTIL_H_
#define _UTIL_H_

#define MALLOC(T) (T *) tlMalloc(sizeof(T))
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <setjmp.h>
#include <time.h>
#include <stdarg.h>

void *tlMalloc(int size);
void *tlRealloc(void *t, int size);
void tlFree(void *t);
void *copy(void *t, int size);
char *copyStr(char *t);
char *catStr(char *s1, char *s2);
long strToLong(char *s);
char *literalStringToString(char *s);
int error(char *msg, ...);
char *getFolder(char *s);
FILE *openFromPath(char *s, char *mode);
char *readFileWithPath(char *path);
char *readFile(FILE *f);

#endif
