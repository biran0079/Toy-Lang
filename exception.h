#ifndef __EXCEPTION_H__
#define __EXCEPTION_H__

#include "execUnit.h"
#include <setjmp.h>
typedef struct Exception {
  jmp_buf buf;
  ExecUnit* finally;
} Exception;

Exception* newException(ExecUnit* finally);
void freeException(Exception* e);

#endif
