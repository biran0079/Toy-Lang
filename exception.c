#include "exception.h"
#include "util.h"

Exception* newException(ExecUnit* finally) {
  Exception* res = MALLOC(Exception);
  res->finally = finally;
  return res;
}

Exception* freeException(Exception* e) {
  freeExecUnit(e->finally);
  free(e);
}
