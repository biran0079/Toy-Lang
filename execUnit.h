#ifndef __EXEC_H__
#define __EXEC_H__

#include "eval.h"
#include "util.h"

typedef struct ExecUnit {
  Value* ev;
  Node* p;
} ExecUnit;

ExecUnit* newExecUnit(Value* ev, Node* p);
void freeExecUnit(ExecUnit* eu);
Value* exec(ExecUnit* eu);

#endif
