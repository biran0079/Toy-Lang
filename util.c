#include "util.h"

void* copy(void* t, int size) {
  void* res = malloc(size);
  memcpy(res, t, size);
  return res;
}

char* copyStr(char* t) {
  return copy(t, strlen(t) + 1);
}

long strToLong(char* s){
  long res = 0;
  while(*s) {
    res*=10;
    res+=*s-'0';
    s++;
  }
  return res;
}

char* literalStringToString(char *s) {
  char *ss = malloc(strlen(s)+1);
  int l = 0;
  s++;  // skil "
  while(*s!='"') {
    if(*s == '\\') {
      s++;
      switch(*s){
        case 'n' : ss[l++] = '\n'; break;
        case 'r' : ss[l++] = '\r'; break;
        case '"' : ss[l++] = '"'; break;
        case '\\' : ss[l++] = '\\'; break;
        default:  ss[l++] = *s; break;
      }
    }else{
      ss[l++] = *s;
    }
    s++;
  }
  ss[l++] = 0;
  return realloc(ss, l);
}

void error(char* format,...) {
  va_list ap;
  va_start(ap, format);
  vfprintf(stderr, format, ap);
  va_end(ap);
  exit(-1);
}
