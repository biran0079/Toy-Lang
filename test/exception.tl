fun f(n){
  try{
    try{
      if(n==10) throw 5;
      if(n%5==0) throw 0;
      if(n%5==1) throw 1;
      if(n%5==2) throw 2;
      if(n%5==3) throw 3;
      if(n%5==4) throw 4;
    }catch(e0){
      if(e0==0) print("e0:", e0);
      else throw e0 ;
    }
  }catch(e1){
    if(e1==1) print("e1:", e1);
    else throw e1;
  }
}

fun g0(n){
  try{
    f(n);
  }catch(e2){
    if(e2==2) print("e2:", e2);  
    else throw e2;
  }
}

fun g1(n){
  try{
    g0(n);  
  }catch(e3){
    if(e3==3) print("e3:", e3);  
    else throw e3;
  }
}

fun g2(n){
  try{
    g1(n);  
  }catch(e4){
    if(e4==4) print("e4:", e4);  
    else throw e4;
  }
}
for(i=0;i<=10;i++){
  try{
    g2(i);
  }catch(e5){
    print("e5:", e5);
  }
}
