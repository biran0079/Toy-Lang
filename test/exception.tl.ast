[stmts, [fun, [id, f], [ids, [id, n]], [stmts, [try, [stmts, [try, [stmts, [if, [==, [id, n], [int, 10]], [throw, [int, 5]]], [if, [==, [%, [id, n], [int, 5]], [int, 0]], [throw, [int, 0]]], [if, [==, [%, [id, n], [int, 5]], [int, 1]], [throw, [int, 1]]], [if, [==, [%, [id, n], [int, 5]], [int, 2]], [throw, [int, 2]]], [if, [==, [%, [id, n], [int, 5]], [int, 3]], [throw, [int, 3]]], [if, [==, [%, [id, n], [int, 5]], [int, 4]], [throw, [int, 4]]]], [id, e0], [stmts, [if, [==, [id, e0], [int, 0]], [call, [id, print], [exps, [string, e0:], [id, e0]]], [throw, [id, e0]]]]]], [id, e1], [stmts, [if, [==, [id, e1], [int, 1]], [call, [id, print], [exps, [string, e1:], [id, e1]]], [throw, [id, e1]]]]]]], [fun, [id, g0], [ids, [id, n]], [stmts, [try, [stmts, [call, [id, f], [exps, [id, n]]]], [id, e2], [stmts, [if, [==, [id, e2], [int, 2]], [call, [id, print], [exps, [string, e2:], [id, e2]]], [throw, [id, e2]]]]]]], [fun, [id, g1], [ids, [id, n]], [stmts, [try, [stmts, [call, [id, g0], [exps, [id, n]]]], [id, e3], [stmts, [if, [==, [id, e3], [int, 3]], [call, [id, print], [exps, [string, e3:], [id, e3]]], [throw, [id, e3]]]]]]], [fun, [id, g2], [ids, [id, n]], [stmts, [try, [stmts, [call, [id, g1], [exps, [id, n]]]], [id, e4], [stmts, [if, [==, [id, e4], [int, 4]], [call, [id, print], [exps, [string, e4:], [id, e4]]], [throw, [id, e4]]]]]]], [for, [exps, [=, [id, i], [int, 0]]], [exps, [<=, [id, i], [int, 10]]], [exps, [++, [id, i]]], [stmts, [try, [stmts, [call, [id, g2], [exps, [id, i]]]], [id, e5], [stmts, [call, [id, print], [exps, [string, e5:], [id, e5]]]]]]]]
