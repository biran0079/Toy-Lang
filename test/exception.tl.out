e0: 0
e1: 1
e2: 2
e3: 3
e4: 4
e0: 0
e1: 1
e2: 2
e3: 3
e4: 4
e5: 5
